src/corehost/cli/hostpolicy.cpp

```diff
@@ -1,279 +1,374 @@
+
+            bool set_app_paths = false;
+
+            // Runtime options config properties.
+            for (int i = 0; i < _hostpolicy_init.cfg_keys.size(); ++i)
+            {
+                // Provide opt-in compatible behavior by using the switch to set APP_PATHS
+                if (pal::cstrcasecmp(_hostpolicy_init.cfg_keys[i].data(), "Microsoft.NETCore.DotNetHostPolicy.SetAppPaths") == 0)
+                {
+                    set_app_paths = (pal::cstrcasecmp(_hostpolicy_init.cfg_values[i].data(), "true") == 0);
+                }
+
+                properties.add(_hostpolicy_init.cfg_keys[i].data(), _hostpolicy_init.cfg_values[i].data());
+            }
+
+            // App paths and App NI paths
+            if (set_app_paths)
+            {
+                properties.add("APP_PATHS", _app_base_cstr.data());
+                properties.add("APP_NI_PATHS", _app_base_cstr.data());
+            }
+
+            // Startup hooks
+            pal::string_t startup_hooks;
+            if (pal::getenv(_X("DOTNET_STARTUP_HOOKS"), &startup_hooks))
+            {
+                pal::pal_clrstring(startup_hooks, &_startup_hooks_cstr);
+                properties.add("STARTUP_HOOKS", _startup_hooks_cstr.data());
+            }
+
+            return StatusCode::Success;
+        }
+
+        const std::unordered_set<pal::string_t>& breadcrumbs() const
         {
-            set_app_paths = (pal::cstrcasecmp(g_init.cfg_values[i].data(), "true") == 0);
+            return _breadcrumbs;
         }
 
-        property_keys.push_back(g_init.cfg_keys[i].data());
-        property_values.push_back(g_init.cfg_values[i].data());
-    }
+    private:
+        hostpolicy_init_t &_hostpolicy_init;
+
+        deps_resolver_t _resolver;
+        const bool _breadcrumbs_enabled;
+        std::unordered_set<pal::string_t> _breadcrumbs;
+
+        // Note: these variables' lifetime should be longer than a call to coreclr_initialize.
+        std::vector<char> _tpa_paths_cstr;
+        std::vector<char> _app_base_cstr;
+        std::vector<char> _native_dirs_cstr;
+        std::vector<char> _resources_dirs_cstr;
+        std::vector<char> _fx_deps;
+        std::vector<char> _app_context_deps;
+        std::vector<char> _clrjit_path_cstr;
+        std::vector<char> _probe_directories;
+        std::vector<char> _clr_library_version;
+        std::vector<char> _startup_hooks_cstr;
+    };
+}
 
-    // App paths and App NI paths
-    if (set_app_paths)
+int run_as_lib(
+    hostpolicy_init_t &hostpolicy_init,
+    const arguments_t &args,
+    std::shared_ptr<coreclr_t> &coreclr)
+{
+    coreclr = g_lib_coreclr.lock();
+    if (coreclr != nullptr)
     {
-        property_keys.push_back("APP_PATHS");
-        property_keys.push_back("APP_NI_PATHS");
-        property_values.push_back(app_base_cstr.data());
-        property_values.push_back(app_base_cstr.data());
+        // [TODO] Validate the current CLR instance is acceptable for this request
```

Comment:
Not sure if this is still needed, in any case we need to deal with the TODO somehow.

Reply:
That is a post preview TODO and somewhat related to the version roll forward logic.

Reply:
OK - maybe create an issue for it - just for tracking :-)